### ISSUE TITLE
ref fields always reference a specific revision?

### ISSUE THREAD

**Issue by zoezoezoe** [webiny/webiny-js#1704]
## This is:

- Bug

## Specifications

  - version: 5.8

## Expected Behavior

Referenced entries in Admin UI should display their latest (published?) revisions.
No warning message should be shown when the latest revision is already published.

## Actual Behavior

Reference entries in Admin UI display the versions they were added when the parent entry was created.
A warning message "Before publishing the main content entry, make sure you publish the following referenced entries" is displayed.

## Steps to Reproduce the Problem

In Admin UI:
1. create content models `Article` and `Author`, where `Article` has a ref field `author` that references the `Author` content type.
2. create an author `Author X` and publish it (revision `#0001`).
3. create an article (`test`) and reference `author` = `Author X`.
4. edit `Author X` so now it reads `Author X2` and publish it (revision `#0002`).
5. go back to article `test` and the author field still shows `Author X`, and there's a warning message "Selected content entry is not published. Make sure to publish it before publishing the main content entry."

Note that Read API returns data from latest revision correctly.

## Detailed Description

It seems like users would expect that the latest (published?) data for a referenced entry is displayed regardless of when the reference was created. As it is right now, it doesn't impact the correctness of the read API, but it would be quite confusing to users.

## Possible Solution

1. store actual `entryId` in `ref.entryId` field, rather than `id`.
2. render the warning message only when the _latest_ revision of the referenced entry is `unpublished`.

**Comment by zoezoezoe** [comment]
Also, the new revision and the warning message would show up in the dropdown only after the web page is reloaded.

**Comment by zoezoezoe** [comment]
Actually, this does introduce a correctness issue in read API: When querying with a `where` clause, we'd need to query by a specific revision, which would leave out other entries that referenced different revisions.

**Comment by Pavel910** [comment]
@zoezoezoe I understand the original issue, it does make sense to always reference _latest published_ record.
What I don't understand is your last comment:
> When querying with a where clause, we'd need to query by a specific revision, which would leave out other entries that referenced different revisions.

Could you provide an actual GQL query to visualize this? Querying _what_ with `where` clause? An example would clarify what you're talkign about. Thanks!

**Comment by zoezoezoe** [comment]
@Pavel910 sure here's an example:
I have content models `Author` and `Article` and `Article.authors` is a ref list of `Author`.
I could have articles referencing different revisions of the same author, say, with ID `A01`.
then I want to have query that gives me the articles written by author `A01` so I'd want to query
`listArticles(where: {authors: "A01"}) { ... }` but right now I can only pass in the exact schema like so:
`listArticles(where: {authors: "A01#0001"}) { ... }`; and this query wouldn't return all other articles that reference a different version of `A01`.

**Comment by Pavel910** [comment]
@zoezoezoe Right, thanks. So, to recap: 
- `ref` field autocomplete in admin UI should always search by _latest published_ (which makes complete sense; and I think it does that, but there's that bug in the UI with message you reported and reference to the original revision)
- and `where` query should search among _any_ revisions if you just send the `entryId` without the revision

Sorry for bothering you for clarifications, but we have _a lot_ of stuff going on, and context switching is a serious problem for us as we often need to jump between completely unrelated user inquiries and stuff we're working on in the background. It's mind blowing to say the least. Clear issues/reports help us understand the request faster and better.

Let me know if the recap is what you're looking for and we'll add it to our product development cycles.

**Comment by zoezoezoe** [comment]
@Pavel910 Yes, that sums up about right. 

Thanks for following up on these reports. No problem at all with more clarification. I'll try to give as much context as possible next time! 

Cheers

**Comment by zoezoezoe** [comment]
This is probably a separate issue but it's related to this (lmk if it belongs to its own issue):

Using examples of `Article` and `Author` above, if I have a query `listArticles` which contain entries that reference *more than one revision* of an `Author`, (e.g. `Article_1.author: author_1#0001` and `Article_2.author: author_1#0002`) I get an error:
```
    {
      "message": "Found non-callable @@iterator",
      "locations": [
        {
          "line": 6,
          "column": 7
        }
      ],
      "path": [
        "listArticles",
        "data",
        1,
        "authors"
      ]
    },
```
The query works if I exclude those that reference different revisions.

This is creating quite a problem for us as we will sometimes update an `Author` when a lot of articles referencing the older revisions have already been published.

**Comment by Pavel910** [comment]
@zoezoezoe nice one! We'll add a test case for it and see if it's reproducible.

**Comment by brunozoric** [comment]
@zoezoezoe
I wrote a test case for the issue and confirmed it's there.
Problem is in our internal DataLoader which batches the entry loading. Basically when loading the entries from the DynamoDB it creates a batch of identical partitionKey/sortKey combinations and it breaks because it is not allowed by the aws-sdk.

Now a question: do you want to load exact referenced revision or latest published?

**Comment by zoezoezoe** [comment]
@brunozoric in our case we would always want to load the latest published revision.

do you have an idea when this could be resolved? until then, we will probably need a workaround to address this issue (the error with the identical keys) in the short term. thanks!

**Comment by brunozoric** [comment]
Hi @zoezoezoe 
I will fix it and it will be out in next release (5.11.0:).
